bootstrap/src/basis/file.mli

```diff
@@ -0,0 +1,122 @@
+open Rudiments
+
+module Path = String
+
+module Mode : sig
+  type t =
+  | R_O   (** open existing file for read, failing if it does not exist *)
+  | W     (** create new or truncate existing file and open for write *)
+  | W_A   (** create new or append to existing file and open for write *)
+  | W_AO  (** append to existing file and open for write, failing if it does not
+              exist *)
+  | W_C   (** create new file and open for write, failing if file exists *)
+  | W_O   (** truncate existing file and open for write, failing if file does
+              not exist *)
+  | RW    (** create new or truncate existing file and open for read and write *)
+  | RW_A  (** create new append to existing file and open for read and write *)
+  | RW_AO (** append to existing file and open for read and write, failing if
+              file does not exist *)
+  | RW_C  (** create new file and open for read and write, failing if file
+              exists *)
+  | RW_O  (** truncate existing file and open for read and write, failing if
+              file does not exist *)
+end
+
+type t
+(** mutable Unix file descriptor *)
+
+type e
+
+val of_path: ?mode:Mode.t -> Path.t -> (t, e) Result.t
+(** [of_path mode path] opens file in [mode] at [path] and returns the resulting
+    file or an error if file could not be opened. *)
+
+val of_path_hlt: ?mode:Mode.t -> Path.t -> t
+(** [of_path mode path] opens file in [mode] at [path] and returns the resulting
+    file or halts if file could not be opened. *)
+
+val read: ?n:usize -> t -> (Bytes.t, e) Result.t
```

Comment:
What do you think about taking a mutable `bytes` parameter here?  Otherwise we have to allocate an array every time we read.  Or maybe we want both APIs, so that we can minimize copying for each case?

Reply:
It depends on what I expect to know from the OS. If we can't know how many bytes the OS has available for us, then we should take a mutable `bytes`. If we were building this on top of poll or epoll, I think that'd be the case. I'll check what kind of info we'll expect from io_uring. If we know how many bytes are available before reading, we could do a better job of constructing small buffers for small reads right before we actually do the read. I need to do a little research on this before I can decide.